### ISSUE TITLE
Ability to set a min language version for a toolchain (maybe also max)

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#16256]
I have a library where I still maintain Java 8 compatibility.
But I want JavaDoc to always be created with at least Java 9 to get the nice search box.
So I could configure a Java 9 toolchain for the JavaDoc task of course.
But if someone builds with Java 11, there is not really a need to discover or auto-provision a Java 9 toolchain.
Because of that it would be nice if you could specify a toolchain with min language version (maybe there are also use-cases for max language version), so that the current toolchain is used if it is recent enough and just another one is searched if it is too old.

Currently you would probably have to do something like
```kts
tasks.javadoc {
    if (!JavaVersion.current().isJava9Compatible) {
        javadocTool.set(javaToolchains.javadocToolFor {
            languageVersion.set(JavaLanguageVersion.of(9))
        })
    }
}

```

**Comment by Vampire** [comment]
While the work-around of course only works if the current version is compatible, a min version and max version would be better as then also another discovered version can be used, like you specify "min java 9", run with java 8 and there is a discovered java 10, then for the JavaDoc task the discovered version can be used without provisioning a new java 9 version.

Maybe even a "preferred version" would then make sense then, so if min version is 9 and 9 and 10 is discovered that the preferred one is used, but that might be overkill.

**Comment by ephemient** [comment]
☕

**Comment by JojOatXGME** [comment]
When you specify Java 17 as the minimum version, I guess Gradle should implicitly specify `-release 17` when it uses a more recent JDK for compiling.

**Comment by 3flex** [comment]
This would be very useful for the JVM daemon's toolchain too. The build might have a minimum required version e.g. 11 and up due to plugin dependencies. There's no need to run it on version 11 exactly.

**Comment by sschuberth** [comment]
> There's no need to run it on version 11 exactly.

Exactly my thoughts. Currently, we potentially unnecessarily bootstrap a toolchain if e.g. Java version 17 is requested, but version 21 is installed.

**Comment by JojOatXGME** [comment]
> Currently, we potentially unnecessarily bootstrap a toolchain

And if you did not configure a plugin to bootstrap new toolchains, the build just fails unnecessarily.

**Comment by ov7a** [comment]
Sorry for the late reply.

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by big-guy** [comment]
We think there's something here.  Toolchains today require an "exact" version to be used, but reality is that a range of versions might be acceptable.

You can compare this to how someone would use dynamic versions + lock files. We should be able to do something like this for toolchains as well.

We're not going to address this before 9.0, but this is something high on the list to consider for 9.x

**Comment by sschuberth** [comment]
> We think there's something here. Toolchains today require an "exact" version to be used, but reality is that a range of versions might be acceptable.

Exactly. IMO the way the `languageVersion` matcher is implemented today simply is too strict. It says "*language* version" anyway, not "JVM version" or so, so the focus should be on compatibility with the *language*.

For people who need strict JVM version matching, https://github.com/gradle/gradle/pull/32522 hopefully brings the solution.
